Add `credentialless` value to COEP (HTML spec)

ArthurSonzogni opened the pull request:
Originally described in: https://github.com/mikewest/credentiallessness

`credentialless` and `require-corp` are similar. One or the other is a requirements for the `window.crossOriginIsolated` capability.
They differ mostly in the fetch specification. `require-corp` requires a CORP header for cross-origin no-cors responses. `credentialless` doesn't, but omits credentials (Cookies, clients certificates, etc...) in the request.

* HTML (https://github.com/whatwg/html/pull/6638)
  * Define how to parse the `credentialless` value.
  * From the HTML spec point of view, `credentialless` and `require-corp` are equivalent. They have been grouped into `compatible with cross-origin isolation` and the HTML spec rewritten to use this concept.

* Fetch: (https://github.com/whatwg/fetch/pull/1229)
  * Define `Cross-Origin-Embedder-Policy allows credentials` algorithm. It omit credentials for no-cors, cross-origin, COEP:credentialless requests.
  * Define `response's` `request-include-credentials` flag.
  * In the `Cross-Origin-Resource-Policy check`, if `embedderPolicy` is `credentialless`, require CORP for navigational responses, and opaque responses with `request-include-credentials`.

See: https://github.com/whatwg/html/issues/6637

----

- [ ] At least two implementers are interested (and none opposed):
   * Chrome: https://chromestatus.com/feature/4918234241302528#details
   * Firefox: https://github.com/mozilla/standards-positions/issues/539  (worth prototyping)
   * Safari: https://lists.webkit.org/pipermail/webkit-dev/2021-June/031898.html (pending)
  
- [X] [Tests](https://github.com/web-platform-tests/wpt) are written and can be reviewed and commented upon at:
   * https://wpt.fyi/results/html/cross-origin-embedder-policy/credentialless

- [X] [Implementation bugs](https://github.com/whatwg/meta/blob/main/MAINTAINERS.md#handling-pull-requests) are filed:
   * Chrome: https://crbug.com/1175099
   * Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1731778
   * Safari: https://bugs.webkit.org/show_bug.cgi?id=230550

(See [WHATWG Working Mode: Changes](https://whatwg.org/working-mode#changes) for more details.)

----

See: https://github.com/whatwg/html/issues/6637


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://whatpr.org/html/6638/browsers.html" title="Last updated on Nov 2, 2021, 12:19 PM UTC (62bc705)">/browsers.html</a>  ( <a href="https://whatpr.org/html/6638/db2d279...62bc705/browsers.html" title="Last updated on Nov 2, 2021, 12:19 PM UTC (62bc705)">diff</a> )
<a href="https://whatpr.org/html/6638/origin.html" title="Last updated on Nov 2, 2021, 12:19 PM UTC (62bc705)">/origin.html</a>  ( <a href="https://whatpr.org/html/6638/db2d279...62bc705/origin.html" title="Last updated on Nov 2, 2021, 12:19 PM UTC (62bc705)">diff</a> )
<a href="https://whatpr.org/html/6638/workers.html" title="Last updated on Nov 2, 2021, 12:19 PM UTC (62bc705)">/workers.html</a>  ( <a href="https://whatpr.org/html/6638/db2d279...62bc705/workers.html" title="Last updated on Nov 2, 2021, 12:19 PM UTC (62bc705)">diff</a> )

ArthurSonzogni commented:
This is still work in progress.

I still need write PR for fetch and ServiceWorker. Those will be more interesting.
Happy to get early feedback if you have in the meantime nevertheless.

@whatwg/cross-origin-isolation @antosart @mikewest @camillelamy @iVanlIsh

ArthurSonzogni commented:
I am happy with the spec.
I wrote page gathering the two PR in a single page if you want to have a broader view:
https://htmlpreview.github.io/?https://github.com/mikewest/credentiallessness/blob/master/index.html

Would you have some suggestion?

domenic commented on source:
```diff
@@ -78591,9 +78591,9 @@ console.assert(iframeWindow.frameElement === null);
-    data-x=""><span>Cross-Origin-Embedder-Policy</span>: <span
-    data-x="coep-require-corp">require-corp</span></code>`.</p></li>
+    <li><p>every <span>Document</span> has a `<code data-x="embedder policy value">
+    Cross-Origin-Embedder-Policy</code>value <span>compatible with crossOriginIsolation</span>.
```
Missing space

ArthurSonzogni replied:
> Also I'm a bit confused on whether we landed on `credentialless` or `cors-or-credentialless`...

It was initially `cors-or-credentialless`, then @annevk suggested switching toward `credentialless` here:
https://github.com/whatwg/html/issues/6637#issuecomment-831113698
`credentialless` is the actual name now.

domenic commented on source:
```diff
@@ -82809,9 +82829,59 @@ interface <dfn>BarProp</dfn> {
    <li><p>Return <var>policy</var>.</p></li>
   </ol>
 
+  <h4>Compatible with crossOriginIsolated</h4>
```
I don't think this needs its own section, and a whole table. Instead it should be under the main "Cross-origin embedder policies value" under the definition of the three values, and it should just be a sentence:

> An [embedder policy value] is **compatible with cross-origin isolation** if it is "`credentialless`" or "`require-corp`".

(Note also that it's spelled "cross-origin isolation", not "crossOriginIsolation".

domenic commented on source:
```diff
@@ -81598,10 +81599,10 @@ interface <dfn>BarProp</dfn> {
-       data-x="coop-same-origin-plus-COEP">same-origin-plus-COEP</code>".</p></li>
+       <li><p>If <var>coep</var>'s <span data-x="embedder-policy-value">value</span> is
+       <span>compatible with crossOriginIsolation</span>, then set <var>policy</var>'s <span
+       data-x="coop-struct-value">value</span> to "<code data-x="coop-same-origin-plus-COEP">
```
Do not introduce linebreaks and whitespace after `>`s; instead you probably want to break after the `<code`.

domenic commented on source:
```diff
@@ -82664,6 +82665,12 @@ interface <dfn>BarProp</dfn> {
+   data-x="">credentialless</code></dfn>"</dt>
+   <dd><p>When this value is used, fetching cross-origin no-CORS resources omits credentials. In
+   exchange, an explicit `<code>Cross-Origin-Resource-Policy</code>` on response is not required.
+   </p></dd>
```
Same line-breaking issue as above.

domenic commented on source:
```diff
@@ -82664,6 +82665,12 @@ interface <dfn>BarProp</dfn> {
+   <dt>"<dfn data-x="coep-credentialless" export for="embedder policy value"><code
+   data-x="">credentialless</code></dfn>"</dt>
+   <dd><p>When this value is used, fetching cross-origin no-CORS resources omits credentials. In
+   exchange, an explicit `<code>Cross-Origin-Resource-Policy</code>` on response is not required.
```
Maybe this should be more general and talk about explicit server permission, to cover CORS as well (like the next bullet pointed does)?

domenic replied:
Sounds good, although "The Other" requests seems a bit confusing. Maybe

> Other requests are sent with credentials, and require the server's....

?

domenic commented on source:
```diff
@@ -82664,6 +82665,12 @@ interface <dfn>BarProp</dfn> {
    without giving explicit permission through the <span>CORS protocol</span> or the
    `<code>Cross-Origin-Resource-Policy</code>` header.</p></dd>
 
+   <dt>"<dfn data-x="coep-credentialless" export for="embedder policy value"><code
```
Nit: I would position this after require-corp since it's easier to understand in contrast to require-corp if one is reading top-to-bottom.

domenic commented on source:
```diff
@@ -82774,18 +82781,31 @@ interface <dfn>BarProp</dfn> {
-    <p>If <var>parsedItem</var> is non-null and <var>parsedItem</var>[0] is "<code
-    data-x="">require-corp</code>":</p>
-
+   <li><p>If <var>parsedItem</var> is non-null</p>
```
Colon before `</p>`, and follow the previous pattern for how the `<p>` is linebreaked and positioned relative to the `<li>`

domenic commented on source:
```diff
@@ -82774,18 +82781,31 @@ interface <dfn>BarProp</dfn> {
-     <li><p>Set <var>policy</var>'s <span data-x="embedder-policy-value">value</span> to "<code
-     data-x="coep-require-corp">require-corp</code>".</p></li>.
+      <!--credentialless-->
+      <li><p>If <var>parsedItem</var>[0] is "<code data-x="">credentialless</code>":</p>
```
Indentation is off for both the `<p>` and `<ol>` here

domenic replied:
No. You have written

```html
<li><p>...

<ol>
 ...
</ol>
```

whereas the prevailing style is

```html
<li>
 <p>...</p>

 <ol>
  ...
 </ol>
</li>
```

I will push this in my fix.

domenic commented on source:
```diff
@@ -82774,18 +82781,31 @@ interface <dfn>BarProp</dfn> {
-     data-x="map exists">exists</span>, then set <var>policy</var>'s <span
-     data-x="embedder-policy-reporting-endpoint">endpoint</span> to <var>parsedItem</var>[1]["<code
-     data-x="coep-report-to">report-to</code>"].</p></li>
+       <li><p>If <var>parsedItem</var>[1]["<code data-x="coep-report-to">report-to</code>"] <span
```
This bullet is duplicated in both branches. It seems like instead this (and the version below) should be something like the following:

4. If _parsedItem_ is non-null and _parsedItem_[0] is "`credentialless`" or "`require-corp`":
    1. Set _policy_'s value to _parsedItem_[0].
    1. If _parsedItem[1]["report-to"] exists, then set _policy_'s endpoint to _parsedItem_[1]["report-to"].

domenic reviewed (commented):
Normatively this is looking good! A number of style and clarity suggestions:

ArthurSonzogni reviewed (commented):
Thanks you Domenic,
I addressed your comments below:

domenic commented on source:
```diff
@@ -78591,9 +78591,9 @@ console.assert(iframeWindow.frameElement === null);
-    data-x=""><span>Cross-Origin-Embedder-Policy</span>: <span
-    data-x="coep-require-corp">require-corp</span></code>`.</p></li>
+    <li><p>every <span>Document</span> has a `<code data-x="embedder policy value">
+    Cross-Origin-Embedder-Policy</code> value <span>compatible with cross-origin isolation</span>.
```
More line-breaking and whitespace problems. Note how adding a newline and spaces before `Cross` makes a visible space in the output, which is undesirable. The same for adding a newline and space before `</p>`.

This reoccurs many times in your PR.

I will push a commit fixing this and other such problems, but please take a look at the diff I upload so that it's clearer in the future how to avoid such problems in the output.

ArthurSonzogni replied:
Hmm, my editor (vim) is able to correctly insert new lines to the right locations, including in the location you fixed if I remove the additional spaces. There must be some unknown (yet) cause it to this in my editor. Next time I should stop relying on tooling and I verify manually every new lines are correct. I didn't realized this. Thanks for the fix!

domenic reviewed (approved):
I pushed a commit with the whitespace fixes; please take a look for next time. Overall this LGTM modulo our discussion on the phrasing of how "credentialless" is explained.

ArthurSonzogni commented:
Thank you!

> Sounds good, although "The Other" requests seems a bit confusing. Maybe

Fixed in f04b0dcd7a980b725e46cded229ec91635047d49

ArthurSonzogni commented:
At the moment, we are evaluating shipping it in Chrome M96 based on the feedback we've got so far. Please let me know if you have any objections.

Until a second web browser is interested implementing it, this will have to stay as two indefinitely open PRs for fetch and HTML + one [monkey](https://wicg.github.io/credentiallessness/)-patch summarizing the two.

annevk commented:
I'm not sure how many other changes are contemplated in this area, but I would be okay with landing this to ease maintenance. However, I do think we should add an implementer warning alongside this value that PNA and ORB are strongly encouraged to be supported in some form.

ArthurSonzogni commented:
I rebased without conflict, and added in the latest commit the warning:

![Capture d’écran du 2021-09-13 14-23-52](https://user-images.githubusercontent.com/4759106/133082930-20d46ca4-1932-4cd0-972e-857821d04426.png)

> I'm not sure how many other changes are contemplated in this area, but I would be okay with landing this to ease maintenance. 

This would encompass this change and the one in Fetch:
- https://github.com/whatwg/html/pull/6638
- https://github.com/whatwg/fetch/pull/1229

annevk commented:
That looks good to me. I have some concerns about the Fetch PR (noted there). The Fetch PR also notes the need for a SW PR which I recall discussing at some point. Is that ready?

ArthurSonzogni commented:
Thanks annevk@. I will take a look at the Fetch PR.

>  The Fetch PR also notes the need for a SW PR which I recall discussing at some point. Is that ready?

This was resolved here:
https://github.com/w3c/ServiceWorker/issues/1592
The issue was about sending responses using CacheStorage/SW in between two environments with different COEP policies.

The resolution was modifying the CORP check in the Fetch specification.
 When the client is using COEP:credentialless, we require CORP when:
- Request made with credentials (<- the important part)
- Navigation requests.

It means, there are no ServiceWorker PR, only a change in Fetch.

annevk commented:
Yeah agreed. We are reluctant and it's conditional upon ORB/PNA working out, but this would ease adoption quite a bit. I do think more pressure could be put on sites to adopt COOP/COEP as-is though.

ArthurSonzogni commented:
> This adds quite a bit of complexity.

`COEP:credentialless` on its own is very lightweight. It can be implemented quickly in 3 steps:
- Conditionally set to false `includeCredential` boolean in the fetch algorithm.
- Add a case in CORP check.
- In many places, replace  "COEP == require-corp" into "COEP == (require-corp | credentialless)".

However, implementing the dependencies: Private Network Access and ORB is indeed a lot of work.

Note that it works independently from anonymous iframe. I don't count this as a dependency. I added a [comment](https://bugs.webkit.org/show_bug.cgi?id=230550) on the webkit bug.

arturjanc commented:
The argument about added complexity is reasonable, but I just wanted to note that there are substantial adoption benefits associated with `COEP:credentiallless`. Namely, while applications can in principle use COEP `require-corp` to enable cross-origin isolation, the requirement to ensure that every loaded resource explicitly opts in (via CORP or CORS) results in a large number of blocking issues that prevent developers from using COEP in most non-trivial applications.

For WebKit specifically, I wonder if the fact that it disables 3p cookies is an advantage in this context. As in, WebKit already sends cross-site requests as effectively credentialless, which aligns very well with the goal here -- this significantly reduces the risk associated with not implementing PNA/ORB. (Though note that same-site requests can still be a concern.)

annevk commented:
No, it does not reduce the risk. The risk was never with CORS requests. The risk of credentialless without PNA/ORB is entirely about resources behind firewalls of sorts.

annevk commented on source:
```diff
@@ -84010,8 +84011,32 @@ interface <dfn interface>BarProp</dfn> {
+   <dt>"<dfn data-x="coep-credentialless" export for="embedder policy value"><code
+   data-x="">credentialless</code></dfn>"</dt>
+   <dd><p>When this value is used, fetching cross-origin no-CORS resources omits credentials. In
+   exchange, and explicit `<code>Cross-Origin-Resource-Policy</code>` is not required. Other
```
```suggestion
   exchange, an explicit `<code>Cross-Origin-Resource-Policy</code>` header is not required. Other
```
(you cannot commit this as-is, as it needs rewrapping and the `source` file is too big to make this feature work, but this was the easiest way to make the comment)

annevk commented on source:
```diff
@@ -84010,8 +84011,32 @@ interface <dfn interface>BarProp</dfn> {
+   data-x="">credentialless</code></dfn>"</dt>
+   <dd><p>When this value is used, fetching cross-origin no-CORS resources omits credentials. In
+   exchange, and explicit `<code>Cross-Origin-Resource-Policy</code>` is not required. Other
+   requests sent with credentials requires the server's explicit permission through the <span>CORS
```
```suggestion
   requests sent with credentials require the server's explicit permission through the <span>CORS
```
(same)

ArthurSonzogni replied:
done

annevk commented on source:
```diff
@@ -84010,8 +84011,32 @@ interface <dfn interface>BarProp</dfn> {
+    <li><a href="https://wicg.github.io/private-network-access/">Private Network Access</a>.
+    <li><a href="https://github.com/annevk/orb">Opaque Response Blocking</a>.
+   </ul>
+   <p>Otherwise, it would let attackers to leverage client's network position to read non public
```
```suggestion
   <p>Otherwise, it would allow attackers to leverage the client's network position to read non-public
```

ArthurSonzogni replied:
Done.

annevk commented on source:
```diff
@@ -84010,8 +84011,32 @@ interface <dfn interface>BarProp</dfn> {
+    <li><a href="https://github.com/annevk/orb">Opaque Response Blocking</a>.
+   </ul>
+   <p>Otherwise, it would let attackers to leverage client's network position to read non public
+   resources by default, using the <span
```
```suggestion
   resources, using the <span
```

ArthurSonzogni replied:
Done.

annevk commented on source:
```diff
@@ -84010,8 +84011,32 @@ interface <dfn interface>BarProp</dfn> {
+  </div>
+
+  <p>An <span>embedder policy value</span> is <dfn>compatible with cross-origin isolation</dfn> if it
+  it either "<code data-x="coep-credentialless">credentialless</code>" or "<code
```
```suggestion
  is "<code data-x="coep-credentialless">credentialless</code>" or "<code
```

annevk commented on source:
```diff
@@ -84117,16 +84142,16 @@ interface <dfn interface>BarProp</dfn> {
    <li>
     <p>If <var>parsedItem</var> is non-null and <var>parsedItem</var>[0] is "<code
-    data-x="">require-corp</code>":</p>
+    data-x="">credentialless</code>" or "<code data-x="">require-corp</code>":</p>
```
Can we use compatible with cross-origin isolation here? Perhaps we should not have "embedder policy value" as a type? And just make it a string. That would also help with the next step.

annevk reviewed (commented):
Overall this looks solid, but I found a number of typos and also have some suggestions for light refactoring.

ArthurSonzogni commented:
Thanks @annevk! I addressed the suggestions from your last review (+ one unresolved). Sorry for the delay, I didn't see your review immediately.

ArthurSonzogni commented:
Thanks Anne! I addressed your comments in 0bb1a99e01483a18586738412f46b57ff98dded9.
Now looking for @domenic for additional comments.

domenic reviewed (approved):
LGTM. Should we merge before or after https://github.com/whatwg/fetch/pull/1229 ?

ArthurSonzogni commented on source:
```diff
@@ -82893,8 +82892,9 @@ interface <dfn interface>BarProp</dfn> {
-    data-x="coop-same-origin">same-origin</span></code>` and `<code
-    data-x=""><span>Cross-Origin-Embedder-Policy</span>: require-corp</code>` together.</p>
+    data-x="coop-same-origin">same-origin</span></code>` and a
+    `<code>Cross-Origin-Embedder-Policy</code>` whose value is <span>compatible with cross-origin
```
Done (Thanks Domenic!)

ArthurSonzogni commented:
Thanks!

> LGTM. Should we merge before or after whatwg/fetch#1229 ?

The fetch specification uses the exported `credentialless` value defined in the HTML specification. As a result, this one should be merged first.
